### PR TITLE
Signup: Add Sign In With Apple button

### PR DESCRIPTION
**Pull request by belcherj**
This pull request is a follow-up of https://github.com/Automattic/wp-calypso/pull/35413 which adds a `Sign In With Apple` button on the `User` step of the `Signup` page:

##### WordPress.com

Before | After
------ | -----
screenshot | screenshot

##### Crowdsignal

Before | After
------ | -----
screenshot | screenshot

##### Jetpack

Before | After
------ | -----
screenshot | screenshot

Note WooCommerce, Akismet and other OAuth clients don't currently display social login buttons, and as such shouldn't have changed. The width of the social login buttons was updated according to the mockups provided.

#### Testing instructions

Do not hesitate to check how the `Signup` page looks when the width of the browser changes:

1. Run `git checkout add/siwa-to-signup` and start your server, or open a [live branch](https://calypso.live/?branch=add/siwa-to-signup)

##### WordPress.com

2. Open the [`Signup` page](http://calypso.localhost:3000/start/user) in an incognito window
3. Assert that the page looks like the screenshot above

##### Crowdsignal

2. Open the Crowdsignal [`Home` page](https://crowdsignal.com/) in an incognito window
3. Click the `CREATE A SURVEY` button
4. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
5. Assert that the page looks like the screenshot above

##### Jetpack

2. [Create](https://jurassic.ninja/create/) a Jetpack site in an incognito window
3. Click the `Set up Jetpack` green button in the `wp-admin` panel of this site
4. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
5. Click the `Create a new account` link in the footer
6. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
7. Assert that the page looks like the screenshot above

##### Other OAuth apps

2. Open the WooCommerce [`Home` page](https://woocommerce.com/) in an incognito window
3. Click the `GET STARTED` link in the header
4. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
5. Assert that no social button is displayed

### CODE REVIEW

File: npm-shrinkwrap.json [hunk 1]
```diff
@@ -6530,7 +6530,8 @@
 						"ansi-regex": {
 							"version": "2.1.1",
 							"resolved": "https://registry.npmjs.org/ansi-regex/-/ansi-regex-2.1.1.tgz",
-							"integrity": "sha1-w7M6te42DYbg5ijwRorn7yfWVN8="
+							"integrity": "sha1-w7M6te42DYbg5ijwRorn7yfWVN8=",
+							"optional": true
```
Comment (Tug) [inline npm-shrinkwrap.json]: I think you're using a different npm version, we shouldn't be updating `npm-shrinkwrap.json` with this PR

File: client/blocks/signup-form/social.jsx [hunk 2]
```diff
@@ -75,7 +74,12 @@ class SocialSignupForm extends Component {
 						responseHandler={ this.handleGoogleResponse }
 						redirectUri={ redirectUri }
 						uxMode={ uxMode }
-						onClick={ this.trackGoogleLogin }
+						onClick={ this.trackSocialLogin.bind( null, 'google' ) }
+					/>
+					<AppleLoginButton
+						clientId={ config( 'apple_oauth_client_id' ) }
+						redirectUri={ `https://${ window.location.host }/start` }
```
Comment (Tug) [inline client/blocks/signup-form/social.jsx]: Let's not use `redirectUri` for now, and only support the popup flow.
Should be pretty close to https://github.com/Automattic/wp-calypso/pull/35428/files#diff-03f7d5aab5df5cd3f255095d092ff252R109

Comment (stephanethomas) [reply client/blocks/signup-form/social.jsx]: Shouldn't we hit the [`handleSocialResponse()` method](https://github.com/Automattic/wp-calypso/blob/19957a61f68c2a2ce24bdd25b0bfa855b6fbc215/client/signup/steps/user/index.jsx#L205) at some point?

File: client/blocks/signup-form/social.jsx [hunk 3]
```diff
@@ -75,7 +84,14 @@ class SocialSignupForm extends Component {
 						responseHandler={ this.handleGoogleResponse }
 						redirectUri={ redirectUri }
 						uxMode={ uxMode }
-						onClick={ this.trackGoogleLogin }
+						onClick={ this.trackSocialLogin.bind( null, 'google' ) }
```
Comment (p-jackson) [inline client/blocks/signup-form/social.jsx]: Binding just the argument and not the `this` feels a little unnatural to me.

What do you think of this?

```suggestion
						onClick={ () => this.trackSocialLogin( 'google' ) }
```

More symbols but still a one-liner :)

File: client/blocks/signup-form/social.jsx [hunk 4]
```diff
@@ -75,7 +84,14 @@ class SocialSignupForm extends Component {
...
+					/>
+
+					<AppleLoginButton
+						clientId={ config( 'apple_oauth_client_id' ) }
+						responseHandler={ this.handleAppleResponse }
+						redirectUri={ null }
+						onClick={ this.trackSocialLogin.bind( null, 'apple' ) }
```
Comment (p-jackson) [inline client/blocks/signup-form/social.jsx]: Same comment as the other use of `.bind`

**Review by p-jackson — COMMENTED** [review]
Apple login all the things! Thanks for the ping :)

I'm not a fan of the way the button width has changed, where's the mockup you refer to? The buttons in the description of #35413 still look full width. Happy to defer to others though, maybe worth a review @monikaburman?

**Review by p-jackson — COMMENTED** [review]
Apple login all the things! Thanks for the ping :)

I'm not a fan of the way the button width has changed, where's the mockup you refer to? The buttons in the description of #35413 still look full width. Happy to defer to others though, maybe worth a review @monikaburman?

**Comment by stephanethomas** [comment]
> I'm not a fan of the way the button width has changed, where's the mockup you refer to? 

You can find the mockup provided by @mattmiklic at c/f9mQpYhh-tr. Let us know what you decide, and will change it 🙂.

File: client/blocks/signup-form/social.jsx (hunk 3 above)
Comment (stephanethomas) [reply client/blocks/signup-form/social.jsx]: I prefer that suggestion as well.

**Comment by mattmiklic** [comment]
@stephanethomas @p-jackson it looks like the "Continue with Google" button has changed since I took the screenshots that I used to base that design on. My intent wasn't to change the button widths; it was actually just to make sure that the Google and Apple buttons were equal width:

Mobile_Web_-_Signup

So let's not make any change to the width of the existing Google button -- just make sure the Apple button matches it.

**Comment by mattmiklic** [comment]
Also just a note fwiw: the text should say "Continue with Apple", not "Sign in with Apple". I'm not sure if the signup buttons are also covered by the text change in #35428.

**Review by stephanethomas — DISMISSED** [review]
When I try to sign up on Crowdsignal with an existing WordPress.com account, I'm redirected to this login screen but the Google button is disabled:

screenshot

I'm getting the following error in the console of my browser:

```
gapi.auth2.ExternallyVisibleError: gapi.auth2 has been initialized with different options. Consider calling gapi.auth2.getAuthInstance() instead of gapi.auth2.init().
```

**Comment by stephanethomas** [comment]
> I'm redirected to this login screen but the Google button is disabled:

Actually simply opening the [`Login` page ](http://calypso.localhost:3000/log-in) of WordPress.com, and clicking the `Sign Up` link in the top right corner will trigger that error @Tug.

**Comment by Tug** [comment]
Yes so unrelated to our PR. I think it might have to do with https://github.com/Automattic/wp-calypso/pull/35019

**Review by p-jackson — COMMENTED** [review]
Thanks for making those changes 👍 This basically LGTM, and everything worked in the testing plan worked as expected when testing on a checked out branch.

I like to be able to follow through and actually sign in with Apple since there was some basic non-cosmetic changes involved in this PR. However the Apple button didn't appear on the live branch (maybe domain doesn't match the one Apple expects? `sign-in-with-apple` is `false` on live branch?).
I also couldn't login with Apple when testing locally. When the popup window closed the login page didn't advance to the next step. Maybe the same domain issue, or because it's not over `https`?

I'm generally happy with the changes though, so if you could share the plan for how/when this gets tested end-to-end then I'd say this is g2g.

**Comment by Tug** [comment]
Yeah this is not really testable on calypso.live as both `sign-in-with-apple` is false and the domain is not an allowed origin.
However, it should work locally, and you should have been able to login/signup. Did you see any error in the console maybe?

**Comment by p-jackson** [comment]
Operator error! Missed this error message before, must have been scrolled out of view:
Screenshot 2019-08-27 at 3 01 56 PM

I was testing what would happen if I told Apple not to give my email to WordPress.com, so I think this is expected.

So then I tried again using a brand new test apple id. This time I'm having a different issue though.
Screenshot 2019-08-27 at 3 01 10 PM
Note that I *have* enabled two-factor on this account, but just using my cell number. Not sure if this is a requirement we've configured as part of our OAuth setup or that Apple OAuth just requires you buy one of their devices before you can use it (!!)

**Comment by p-jackson** [comment]
Ok I see that there's a difference between two-*factor* and two-*step* now. Sorry for the churn.

I have been able to log in using Apple now, but it only seems to work half the time. When it fails there isn't any error in the UI or in the console. I can't seem to find any sort of pattern for when it fails unfortunately.

@Tug this might be an issue unrelated to this PR, and since this is still behind a feature flag you might want to merge anyway. Up to you.

**Comment by simison** [comment]
✔️ Signing up with Apple using "don't reveal my email" worked great, apart from website address confusion. Note how my email is now `@privaterelay.appleid.com`
Screenshot 2019-08-27 at 09 34 41

✔️ Logging in with that same apple ID logs me back to that same account.

✔️ Connecting my a8c account to apple id failed due error. Seems expected?
Screenshot 2019-08-27 at 09 45 36


🚩 When creating a new Jetpack site (using https://jurassic.ninja/) and connecting it using Calypso-Live branch, I see white screen and some failures in console:

image

You can test locally by appending `&calypso_env=local` to the Jetpack connection URL to get to localhost version. There's some value for live sites but I cannot remember what exactly (cc @sirreal you might know?)

**Comment by Tug** [comment]
> I have been able to log in using Apple now, but it only seems to work half the time. When it fails there isn't any error in the UI or in the console. I can't seem to find any sort of pattern for when it fails unfortunately.

@stephanethomas seems to have experienced the same trouble shortly after activating 2FA. I was not able to reproduce and I don't think we can do anything about it but it's a bit concerning.

> 🚩 When creating a new Jetpack site (using https://jurassic.ninja/) and connecting it using Calypso-Live branch, I see a white screen and some failures in the console:

Those seems to be google client errors and I could bet it also happens on `master`.

**Comment by simison** [comment]
> > When creating a new Jetpack site (using https://jurassic.ninja/) and connecting it using Calypso-Live branch, I see a white screen and some failures in the console:
>
> Those seems to be google client errors and I could bet it also happens on master. Sorry about that, testing instructions should be updated

Hmm now I re-tested and all worked fine (both `master` and the live branch):

Screenshot 2019-08-28 at 10 27 22

Only saw this in the console:

Screenshot 2019-08-28 at 10 56 24

**Comment by Tug** [comment]
> Only saw this in the console:

Yep, that's also an error from the google js client, when the origin is not allowed.

**Pull request merged**